app/models/course.rb

```diff
@@ -670,7 +671,13 @@ def is_published?
   end
 
   def has_unpublished_changes?
```

Comment:
I know this one's painful! 

Does it imply `published, but with unpublished changes`? or just that there are unpublished changes

Would `(published? && draft_enrichment.present?)` be true if there were an old draft that was never published followed by a later published enrichment? Should this return true in that case?

(This would also cause `all_enrichments_are_published?` not to be true)

then in `(last_published_at.present? && enrichment_not_withdrawn?)`, we're assuming that the potentially-withdrawn enrichment is the one that was published. I think i now see why we look at the "current" enrichment in `last_published_at`. This is confusing.

Does this method actually want to check whether there exists a draft enrichment created after the last time the course was published?

Reply:
Oh yes the source of all pain. Both... well, it needs to cover multiple scenarios due to `app/views/publish/courses/_course_button_panel.html.erb:2`. 

I assume it is possible we could end up with a draft enrichment that is superseded, but my instinct is that this is an edge case, then again I wouldn't be too surprised!